Optimise tags use in `Chunk` implementations

guizmaii opened the pull request.

guizmaii committed:
Optimise tags use in `Chunk` implementations

guizmaii commented on core/shared/src/main/scala/zio/Chunk.scala:
```diff
@@ -2470,6 +2471,9 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
       extends Arr[A]
       with ChunkIterator[A] { self =>
+
+    val classTag: ClassTag[A] = Tags.fromValue(array(0))
```
The previous implementation was inherited from `Arr` and was:
```scala
implicit val classTag: ClassTag[A] =
      ClassTag(array.getClass.getComponentType)
```

See https://github.com/zio/zio/pull/10079/files#diff-692b5e51eb20ff6846be420c52c1b9830e2f756b69e280e48234410676013254L1597-L1598

Not sure which one to prefer

guizmaii commented on core/shared/src/main/scala/zio/Chunk.scala:
```diff
@@ -1594,9 +1602,6 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
 
     val array: Array[A]
 
-    implicit val classTag: ClassTag[A] =
-      ClassTag(array.getClass.getComponentType)
```
This can be a constant in most `Arr` implementations, which allows us to remove this `val` in most implementation (all except `AnyRefArray` actually)

guizmaii committed:
Make it biased towards Scala3

guizmaii commented on core/shared/src/main/scala/zio/Chunk.scala:
```diff
@@ -1538,9 +1548,10 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
 
-    implicit val classTag: ClassTag[A] = Chunk.classTagOf(chunk)
+    @threadUnsafe
+    lazy val classTag: ClassTag[A] = Chunk.classTagOf(chunk)
```
Here, I made some choices that can be discussed:
1. Use a `lazy val` instead of a `val` to avoid calculating them when unnecessary. 
  The goal being to make the Chunk implementations as lightweight as possible in terms of memory when instantiated, as so as fast as possible to instantiate
2. Biased it towards Scala3 by using `@threadUnsafe` (only available in Scala3) to make it as performant as possible (but in Scala2, it'll be less performant)

hearnadam replied:
I don't think this should be a lazy val. The field will still be present in the bytecode, right? And for most users (Scala 2), you will see voilatile reads and locking, right?

guizmaii replied:
> And for most users (Scala 2), you will see voilatile reads and locking, right?

You assume that most of our users are using Scala2. 
I assume that it'll be fewer and fewer; if so, I think that our effort should be biased and focus towards Scala3. 
Again, it's a choice (not a decision) I made consciously. I'm ok to hear people disagreeing and ok to change to something more Scala2-friendly.

Maybe making it a `def` would be good enough? 🤔 

> The field will still be present in the bytecode, right?

That's a good question. I need to double-check that

guizmaii replied:
@hearnadam I decided to change my mind and I applied your recommendations. They're not `val`s 
Thanks for your review! 🙂

guizmaii commented on core/shared/src/main/scala/zio/Chunk.scala:
```diff
@@ -2421,8 +2430,7 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
 
-    implicit val classTag: ClassTag[T] =
-      ops.classTag
+    def classTag: ClassTag[T] = ops.classTag
```
`ops.classTag` is always a constant

guizmaii commented on core/shared/src/main/scala/zio/Chunk.scala:
```diff
@@ -2470,6 +2478,10 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
       with ChunkIterator[A] { self =>
+
+    @threadUnsafe
+    lazy val classTag: ClassTag[A] = Tags.fromValue(array(0))
```
The previous implementation was inherited from `Arr` and was:
```scala
    implicit val classTag: ClassTag[A] =
      ClassTag(array.getClass.getComponentType)
```

This new implemetation is inspired from the one in `VectorChunk`

Not sure which one to prefer

guizmaii committed:
Merge branch 'series/2.x' into ligher_chunk_tags

guizmaii committed:
Make the Scala2 `threadUnsafe` annotation stub private to ZIO to avoid polluting users code

guizmaii commented on core/shared/src/main/scala-2/zio/Scala3Annotations.scala:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Stubs for annotations that exist in Scala 3 but not in Scala 2
+ */
+private[zio] object Scala3Annotations {
```
Trick copied from Caliban

kyri-petrou replied:
I can't help it but feel a bit honoured when others start using my hacks 😆

hearnadam commented on core/shared/src/main/scala/zio/Chunk.scala:
```diff
@@ -1890,8 +1900,8 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
-    implicit val classTag: ClassTag[A] =
-      Tags.fromValue(vector(0))
+    @threadUnsafe
+    lazy val classTag: ClassTag[A] = Tags.fromValue(vector(0))
```
I think we should just make this a regular val throughout.

guizmaii replied:
Done

guizmaii committed:
Review: Prefer `val` over `lazy val`

kyri-petrou commented on core/shared/src/main/scala/zio/Chunk.scala:
```diff
@@ -1412,13 +1412,21 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
    */
   private[zio] def classTagOf[A](chunk: Chunk[A]): ClassTag[A] =
-    chunk match {
+    (chunk: @switch) match {
```
Does `@switch` work here or does the compiler emit a warning that it could not emit a switch statement?

guizmaii replied:
Good catch! Removed `@switch` — type pattern matching uses `instanceof` checks, not a jump table, so the annotation was ineffective here.

kyri-petrou commented on core/shared/src/main/scala/zio/Chunk.scala:
```diff
@@ -1412,13 +1412,21 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
   private[zio] def classTagOf[A](chunk: Chunk[A]): ClassTag[A] =
-    chunk match {
+    (chunk: @switch) match {
+      case Empty                    => classTag[java.lang.Object].asInstanceOf[ClassTag[A]]
```
I think it's better to avoid the equality check and use `_: Empty.type` instead

guizmaii replied:
Good point! Changed to `_: Empty.type` to use an `instanceof` check instead of `equals()`. Cheaper and more semantically correct.

hearnadam committed:
Merge branch 'series/2.x' into ligher_chunk_tags

guizmaii committed:
Merge branch 'series/2.x' into ligher_chunk_tags

guizmaii committed:
Review: Remove @switch and use instanceof check for Empty in classTagOf

- Remove @switch annotation since type pattern matching uses instanceof
  checks, not a jump table, making the annotation ineffective
- Use `_: Empty.type` instead of `Empty` to avoid equals() overhead
  and use a cheaper instanceof check instead

guizmaii commented:
@kyri-petrou I took your comments into account. Ready to review. Sorry for the delay

guizmaii committed:
Use ClassTag.Object constant instead of classTag[java.lang.Object]
